book/animations.md

```diff
@@ -5,56 +5,153 @@ prev: scheduling
+animation lets us request that some JavaScript run on the next
+*frame*, and we can have that JavaScript change the page slightly.
+To do this repeatedly, we'll need code like this:
+
+``` {.javascript file=example-opacity-js}
+function run_animation_frame() {
+    if (animate())
+        requestAnimationFrame(run_animation_frame);
+}
+requestAnimationFrame(run_animation_frame);
+```
 
-[^video-anim]: And video-like animations include videos, animated images, and
-animated canvases.
+Here `animate` makes some small change to the page to give the
+impression of continuous change. It returns `true` until it's done
+animating, and then stops. By changing what `animate` does we can
+change what animation occurs.
 
-[^excuse]: Video animations is a topic unto itself, but is beyond the
-scope of this book. Arguably, canvas is a bit different than the other two,
-since it's implemented by developer scripts. And of course a canvas can have
-animations within it that look about the same as some DOM animations.
+Let's write a fade animation. We can fade in something out by smoothly
+transitioning its `opacity` value from 0.0 to 0.999.[^why-not-one] If we
+want to do this animation over 120 frames (about two seconds), that
+means we need to increase the opacity by about 0.008 on each frame.
 
-DOM and input-driven animations can be sub-categorized into *layout-inducing*
-and *visual*. An animation is layout-inducing if it changes an input to layout;
-animating `width` is one example that we'll encounter in this chapter.
-Otherwise the animation is visual, such as animations of `opacity` or
-`background-color`.
+[^why-not-one]: Real browsers apply certain optimizations when opacity
+is exactly 1, so real-world websites often start animations at 0.999.
+That way, the animation is smooth. So it's easier to dig into the
+performance of this example on a real browser with 0.999 opacity.
+Starting animations at 0.999 is also a common trick used on web sites
+that want to avoid visual popping of the content as it goes in and out
+of GPU-accelerated mode. I chose 0.999 because the visual difference
+from 1.0 is imperceptible.
+
+For example, let's animate this `div` containing the word "Test":
+
+``` {.html file=example-opacity-html}
+<div>This text fades out</div>
+```
+
+The `animate` function will track how many frames have occurred and 
+
+``` {.javascript file=example-opacity-js}
+var div = document.querySelectorAll("div")[0];
+var total_frames = 120;
+var current_frame = 0;
+var change_per_frame = 0.999 / total_frames;
+function animate() {
+    current_frame++;
+    var new_opacity = current_frame * change_per_frame;
+    div.style = "opacity:" + new_opacity;
+    return current_frame < total_frames;
+}
+```
 
+Here's how it looks; you'll probably need to refresh the page or [open
+it full-screen](examples/example13-opacity-raf.html) to watch the
+animation from the beginning.
 
+<iframe src="examples/example13-opacity-raf.html"></iframe>
```

Comment:
I do think it's important for this to keep running forever and not force a reload of the page to restart. Otherwise it'll be too easy to miss it while reading.

Reply:
Maybe it would be enough to have the iframe page go back and forth, and add sentence saying "here's a demo of it fading out, then fading back in and repeating"

Reply:
What if instead of fading back and forth the animation just restarted at transparent? We could even just trigger an iframe refresh. (Mainly, the direction-switching code seemed complicated to me.)

Reply:
The problem is that once the demo is ported to a css animation in the next section, I wanted it to look identical. Which is why it's an easing function in each direction.

Reply:
Aha! Got it! What if instead of a fully-automated animation that cycles, we had one with buttons to animate one way or the other? That might be better for readers (no distracting animation if they don't want it) and works for both cases.

Reply:
ok sure, if you're up for that.

Reply:
Ok, I added buttons. It works great in our browser!

Reply:
\o/